### PR TITLE
IGAPP-1234: Support multipois WEB

### PR DESCRIPTION
**Pull request by sarahsporck**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

What's new? 
- properly restore scroll position feels much smoother now
- filtered list when selecting a multipoi on the map (check out Cafè Tür an Tür)
- new multipoi icon (this one could be improved, but I think I need @f1sh1918 help for that)

What's changed?
- responsibilities are now split correctly (the map handles flyTo, not the pois page)
- platform specific code has moved to PoisMobile/PoisDesktop correspondingly, improving maintainability hopefully
- removed settimeout from flyTo (will close IGAPP-1154)

What to test?
- web small viewport and large viewport
- restore scroll position when navigating from details to list
- multipoi selection/deselection on map and in list
- whether still flyTo works as before

### CODE REVIEW

File: web/src/hooks/usePoiFeatures.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import { Map } from 'maplibre-gl'
+import { useCallback, useEffect, useMemo } from 'react'
+import { useNavigate, useSearchParams } from 'react-router-dom'
+
+import { GeoJsonPoi, isMultipoi, PoiFeature, PoiModel } from 'api-client'
+
+const multipoiKey = 'multipoi'
+
+const usePoiFeatures = (
```
Comment (steffenkleinle) [inline web/src/hooks/usePoiFeatures.ts:9]: Optional: I struggled a bit with the responsibilities between the two hooks, PoisPage and PoisMobile/Desktop. I created a proposal which might make things a little more clearer: #956
Feel free to reject.

**Review by steffenkleinle — COMMENTED** [review]
Amazing, I love the refactoring! :heart_eyes: Good job :rocket: 

Tested on firefox, both desktop and mobile.

One observation: If I click next to a poi that is part of a multipoi, I think I would expect to completely deselect it (poi and multipoi) and go back to the full list, but currently this is not the case as it only deselects the poi and the multipoi is still selected.
Not sure if my expectation is just wrong or if this should actually be changed. At least it is not a big issue.

File: web/src/components/PoiPanelNavigation.tsx [hunk 2]
```diff
@@ -42,31 +42,31 @@ const Icon = styled.img<{ direction: string }>`
 `
 
 type PoiPanelNavigationProps = {
-  switchFeature: (step: 1 | -1) => void
+  switchPoi: (step: 1 | -1) => void
```
[the remark this answers is not in the document]
Comment (sarahsporck) [reply web/src/components/PoiPanelNavigation.tsx:45]: I tried naming it poi for `PoiModel` and poiFeature for `GeoJsonPoi` and finally feature for `PoiFeature`

File: web/src/components/MapView.tsx [hunk 3]
```diff
@@ -31,98 +32,96 @@ const MapContainer = styled.div`
...
       event.originalEvent.stopPropagation()
       const feature = event.features && (event.features[0] as unknown as PoiFeature)
       if (feature) {
-        selectFeature(feature, false)
-        changeSnapPoint(1)
+        selectFeature(
+          {
+            ...feature,
+            properties: {
+              pois: JSON.parse(feature.properties.pois as unknown as string),
```
[the remark this answers is not in the document]
Comment (sarahsporck) [reply web/src/components/MapView.tsx:94]: No. Apparently the library struggles to parse arrays. https://github.com/maplibre/maplibre-gl-js/issues/1325
I'll add the link.

File: web/src/hooks/usePoiFeatures.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        mapRef.stop()
+      }
+      if (!newFeatureOnMap) {
+        deselectFeature()
+      } else if (isMultipoi(newFeatureOnMap)) {
+        navigate(`.?${new URLSearchParams([[multipoiKey, newFeatureOnMap.id as string]])}`)
+      } else {
+        const newPoiFeature = newFeatureOnMap.properties.pois[0]
+        if (newPoiFeature && currentFeatureOnMap?.properties.pois.includes(newPoiFeature)) {
+          navigate(`${newPoiFeature.slug}?${searchParams}`)
```
[the remark this answers is not in the document]
Comment (sarahsporck) [reply web/src/hooks/usePoiFeatures.ts:62]: No, it is not. See issue description.

File: web/src/hooks/usePoiFeatures.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      setSearchParams(new URLSearchParams())
+    }
+  }, [currentFeatureOnMap?.properties.pois, currentPoi, setSearchParams])
+  return { currentFeatureOnMap, currentPoi }
+}
+
+export const usePoiHandles = (
+  currentFeatureOnMap: PoiFeature | null,
+  currentPoi: PoiModel | null,
+  mapRef?: Map
```
[the remark this answers is not in the document]
Comment (sarahsporck) [reply web/src/hooks/usePoiFeatures.ts]: and actually it also seems that this is not necessary. will remove it.

File: web/src/hooks/usePoiFeatures.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply web/src/hooks/usePoiFeatures.ts:57]: Then we should kinda not need this check right?

File: web/src/components/__tests__/PoiListItem.spec.tsx [hunk 6]
```diff
@@ -9,23 +9,24 @@ import PoiListItem from '../PoiListItem'
...
+    const { getByText } = renderWithTheme(<PoiListItem selectPoi={selectPoi} poi={poiFeature} />)
 
-    expect(getByText(feature.properties.title)).toBeTruthy()
+    expect(getByText(poiFeature.title)).toBeTruthy()
     expect(getByText('pois:distanceKilometre')).toBeTruthy()
-    expect(getByText(feature.properties.category!)).toBeTruthy()
+    expect(getByText(poiFeature.category!)).toBeTruthy()
   })
 
   it('should select feature', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/PoiListItem.spec.tsx:25]:
```suggestion
  it('should select poi', () => {
```

File: web/src/components/__tests__/PoisDesktop.spec.tsx [hunk 7]
```diff
@@ -1,65 +1,91 @@
...
 import { renderWithRouterAndTheme } from '../../testing/render'
 import PoisDesktop from '../PoisDesktop'
 
 jest.mock('react-i18next')
+jest.mock('../MapView', () => {
+  const { forwardRef } = jest.requireActual('react')
+  return {
+    __esModule: true,
+    default: forwardRef((_: object, ref: React.Ref<HTMLDivElement>) => <div ref={ref}>MapView</div>),
+  }
```
Comment (LeandraH) [inline web/src/components/__tests__/PoisDesktop.spec.tsx:22]: 😬

File: web/src/components/__tests__/PoisMobile.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  it('should render filtered poiList & toolbar components for multipoi feature', () => {
+    const multipoiFeature = features.find(feature => feature.properties.pois.length > 1)!
+    const { queryByText } = renderPoisDesktop(undefined, multipoiFeature)
+
+    expect(queryByText('detailsHeader')).toBeFalsy() // because bottomsheet is not fullscreen
+    expect(queryByText('listTitle')).toBeFalsy()
+    expectPoiList(queryByText, multipoiFeature.properties.pois)
+  })
+
+  it('should render poiList & toolbar components no poi is provided', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/PoisMobile.spec.tsx:83]:
```suggestion
  it('should render poiList & toolbar components if no poi is provided', () => {
```

File: web/src/components/__tests__/PoisDesktop.spec.tsx [hunk 9]
```diff
@@ -1,65 +1,91 @@
...
-    expect(queryByText(feature.properties.category!)).toBeNull()
-    expect(queryByText(poi.content)).toBeNull()
+    expect(queryByText('detailsHeader')).toBeTruthy()
+    expect(queryByText('listTitle')).toBeFalsy()
+    expectPoiList(queryByText, multipoiFeature.properties.pois)
   })
 
-  it('should not render toolbar in detail view if showNavigation is set to false', () => {
-    const { queryByText, queryByLabelText } = renderPoisDesktop(false, feature)
+  it('should render poiList & toolbar components no poi is provided', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/PoisDesktop.spec.tsx:85]:
```suggestion
  it('should render poiList & toolbar components if no poi is provided', () => {
```

**Review by LeandraH — APPROVED** [review]
Nice! Tested on Firefox. I agree with Steffen, the back function from a selected POI back to the multi POI is not quite intuitive but I think it's fine.

**Comment by steffenkleinle** [comment]
Another thing I just realized: If I go to the map (http://localhost:9000/testumgebung/de/locations) and then press in the list on `Cafe Tür an Tür`, it flies all the way to the cafe, jumps back to berlin and then flies to the cafe again.

**Review by f1sh1918 — COMMENTED** [review]
Really nice refactoring!
There is one issue with the fullscreen bottom action sheet
The marked back element should only be displayed on fullscreen. Now there are two buttons with the same function.
image

**Comment by sarahsporck** [comment]
>     1. There is one issue with the fullscreen bottom action sheet
>        The marked back element should only be displayed on fullscreen. Now there are two buttons with the same function.

Nice find. Somehow I didn't properly check whether the bottom sheet was full screen.

> 
>     2. Clicking on map should deselect the multipoi

Hmm. Not 100% sure what you refer to here. What means deselect in this context? Say I have selected a multipoi and then in this multipoi I have selected a poi. What should happen if i click on the map?

**Comment by steffenkleinle** [comment]
> > 2. Clicking on map should deselect the multipoi
> 
> Hmm. Not 100% sure what you refer to here. What means deselect in this context? Say I have selected a multipoi and then in this multipoi I have selected a poi. What should happen if i click on the map?

My personal expectation that after clicking on the map, nothing is selected anymore, no matter if before a poi, multipoi or a multipoi AND a poi were selected. So a click clears the whole selection.


Also just to make sure that you did not miss my comment:
> Another thing I just realized: If I go to the map (http://localhost:9000/testumgebung/de/locations) and then press in the list on `Cafe Tür an Tür`, it flies all the way to the cafe, jumps back to berlin and then flies to the cafe again.

File: web/src/hooks/useFeatureLocations.ts [hunk 10]
```diff
@@ -1,23 +1,33 @@
...
 type FeatureLocationsType = { features: PoiFeature[]; pois: PoiModel[] }
 
-const useFeatureLocations = (cityCode: string, languageCode: string): ReturnType<FeatureLocationsType> => {
+const useFeatureLocations = (
+  cityCode: string,
+  languageCode: string,
+  coordinates?: LocationType
+): ReturnType<FeatureLocationsType> => {
   const requestFeatureLocations = useCallback(async () => {
-    const { coordinates } = await getUserLocation()
```
Comment (steffenkleinle) [inline web/src/hooks/useFeatureLocations.ts:12]: Why did you remove this from here?

File: web/src/components/MapView.tsx [hunk 11]
```diff
@@ -37,35 +31,23 @@ type MapViewProps = {
...
   featureCollection,
   selectFeature,
   changeSnapPoint,
   currentFeature,
-  cityModel,
+  viewport,
+  setViewport,
   children,
 }: MapViewProps): ReactElement => {
-  const bboxViewport = useMemo(
```
Comment (steffenkleinle) [inline web/src/components/MapView.tsx]: Why did you move the viewport logic?

File: web/src/components/PoisMobile.tsx [hunk 12]
```diff
@@ -160,11 +172,16 @@ const PoisMobile = ({
...
+            direction={direction}
+            viewportSmall
+            text={t('detailsHeader')}
+          />
         )}
         <ListContainer>
-          {currentPoi && currentPoiFeature ? (
-            <PoiDetails poi={currentPoi} feature={currentPoiFeature} direction={direction} t={t} />
+          {currentPoi ? (
+            <PoiDetails poi={currentPoi} feature={currentPoi.getFeature(userLocation)} direction={direction} t={t} />
```
Comment (steffenkleinle) [inline web/src/components/PoisMobile.tsx:184]: If this is the reason why you did all the changes, the following should still work: 
```
const currentPoiFeature = currentFeatureOnMap?.properties.pois.find(poi => poi.slug === currentPoi?.slug)
```

Not sure if that includes the user location though.

File: web/src/components/MapView.tsx (hunk 11 above)
Comment (sarahsporck) [reply web/src/components/MapView.tsx]: I thought about the situation where one reduces the size of his/her browser window or switches between landscape and horizontal on his/her tablet. Then the viewport will be reset, which does not feel right.
I also thought about moving reverting most of the refactoring and putting the map back into the pois page again, but I still think that generally the current implementation makes more sense.

File: web/src/hooks/useFeatureLocations.ts (hunk 10 above)
Comment (sarahsporck) [reply web/src/hooks/useFeatureLocations.ts:12]: To dynamically calculate the distance for each feature. (`currentPoi.getFeature(coordinates)`)

**Comment by sarahsporck** [comment]
> Another thing I just realized: If I go to the map (http://localhost:9000/testumgebung/de/locations) and then press in the list on `Cafe Tür an Tür`, it flies all the way to the cafe, jumps back to berlin and then flies to the cafe again.

Could not reproduce this so far. What browser are you using?

File: web/src/components/MapView.tsx (hunk 11 above)
Comment (steffenkleinle) [reply web/src/components/MapView.tsx]: I think it is nice the way it is :)

File: web/src/hooks/usePoiFeatures.ts [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+          : features.find(feature => feature.properties.pois.some(poi => poi.slug === slug)),
+      [features, searchParams, slug]
+    ) ?? null
+
+  useEffect(() => {
+    if (currentPoi && !currentFeatureOnMap?.properties.pois.some(poiFeature => poiFeature.slug === currentPoi.slug)) {
+      // remove multipoi from search params if current poi is not part of this multipoi
+      setSearchParams(new URLSearchParams())
+    }
+  }, [currentFeatureOnMap?.properties.pois, currentPoi, setSearchParams])
```
Comment (steffenkleinle) [inline web/src/hooks/usePoiFeatures.ts:38]: So I tested this again by uncommenting and I did not find a case where this is necessary, everything worked just fine without. Do you have any more information on how to reproduce this edge case for which this is necessary?

File: web/src/components/PoiGoBack.tsx [hunk 14]
```diff
@@ -0,0 +1,74 @@
+import React, { memo, ReactElement } from 'react'
```
Comment (steffenkleinle) [inline web/src/components/PoiGoBack.tsx:1]: Could you still rename this to just `GoBack` or similar?

**Review by steffenkleinle — COMMENTED** [review]
I can't reproduce the jumping back to Berlin anymore. However,the app still reloads something and flickers quickly sometimes.
I think this might happen if the user location becomes available during flyTo. Maybe we can still pass the old data without user location while loading them again? Like we should not trigger a reload from the cms if we get coordinates.

File: web/src/hooks/usePoiFeatures.ts (hunk 13 above)
Comment (sarahsporck) [reply web/src/hooks/usePoiFeatures.ts:38]: So I think I added this to keep a consistent state and not display sth. weird. Try navigating to `http://localhost:9000/testumgebung/en/locations/rathausplatz?multipoi=4`. Imo this url should not be valid. Therefore I added this useEffect.

Comment (f1sh1918) [reply web/src/hooks/usePoiFeatures.ts:38]: hm for me also everything works fine @sarahsporck 
I go to the url and the camera goes to the multipoi. I can't see any difference on chrome nor firefox with or without that. I don't think this is needed as a hook

Comment (sarahsporck) [reply web/src/hooks/usePoiFeatures.ts:38]: So when you go to the given url, without this `useEffect` you should see that the multipoi is selected on the map and the camera also moves to the multipoi, but in the list `Rathausplatz` is selected. Please tell me I don't go crazy here ^^

Comment (f1sh1918) [reply web/src/hooks/usePoiFeatures.ts:38]: you are right @sarahsporck :)

**Review by f1sh1918 — APPROVED** [review]
Works nice :) tested on chrome
